Add rsync-args flag to oc rsync command

csrwng opened the pull request:
Allows additional flags to be passed to rsync:

```
--include
--exclude
--progress
--no-perms
```

displays a warning if attempting to use these flags with the tar strategy.
Fixes #6143

csrwng commented:
@smarterclayton or @bparees ptal

smarterclayton commented on pkg/cmd/cli/cmd/rsync/rsyncargs.go:
```diff
@@ -0,0 +1,196 @@
+//    other -> inDoubleQuote (accmulate char for arg)
+// inEscape
+//    any -> nextState (accumulate char for arg)
+func parseArgs(str string) ([]string, error) {
```
Yuck

csrwng replied:
I found a couple of go libraries that did this (with caveats)
https://github.com/mattn/go-shellwords - no license
https://github.com/mgutz/str - panics on errors

I need to parse the different string args so I can pass them to the go cmd.exec ... alternatively, we could just white-list the args we accept and allow you to specify them directly on the command.

smarterclayton replied:
Sigh...

On Wed, Dec 16, 2015 at 3:37 PM, Cesar Wong <notifications@github.com>
wrote:

> In pkg/cmd/cli/cmd/rsync/rsyncargs.go
> <https://github.com/openshift/origin/pull/6351#discussion_r47827380>:
>
> > +//    other -> inArgument (acccumulate char for arg)
> > +// inSingleQuote
> > +//    space -> inSingleQuote (accumulate char for arg)
> > +//    single-quote -> inArgument
> > +//    double-quote -> inSingleQuote (accumulate char for arg)
> > +//    escape-char -> inSingleQuote (accumulate char for arg)
> > +//    other -> inSingleQuote (accumulate char for arg)
> > +// inDoubleQuote
> > +//    space -> inDoubleQuote (accmulate char for arg)
> > +//    single-quote -> inDoubleQuote (accmulate char for arg)
> > +//    double-quote -> inArgument
> > +//    escape-char -> inEscape (nextState = inDoubleQuote)
> > +//    other -> inDoubleQuote (accmulate char for arg)
> > +// inEscape
> > +//    any -> nextState (accumulate char for arg)
> > +func parseArgs(str string) ([]string, error) {
>
> I found a couple of go libraries that did this (with caveats)
> https://github.com/mattn/go-shellwords - no license
> https://github.com/mgutz/str - panics on errors
>
> I need to parse the different string args so I can pass them to the go
> cmd.exec ... alternatively, we could just white-list the args we accept and
> allow you to specify them directly on the command.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6351/files#r47827380>.
>

bparees replied:
there's a certain rationale to only supporting the minimal args we care about and doing it in a first class way, and then telling people who want the full range of rsync commands to just oc exec rsync themselves.

liggitt replied:
this makes me really nervous...

csrwng replied:
> this makes me really nervous...

which part :) the parsing of stuff ourselves? or passing more args to rsync?

bparees replied:
i assume you mean "writing our own argument parser" makes you nervous and not the suggestion to wrapper a subset of first class rsync arguments.

liggitt replied:
yes, custom arg parsing

bparees commented on pkg/cmd/cli/cmd/rsync/copyrsync.go:
```diff
@@ -61,6 +61,7 @@ func newRsyncStrategy(f *clientcmd.Factory, c *cobra.Command, o *RsyncOptions) (
 	if o.Delete {
 		flags = append(flags, "--delete")
 	}
+	flags = append(flags, o.RsyncArgs...)
```
this fullfills the TODO on line 51 so you can remove it... right?

csrwng replied:
true, yes

bparees commented on pkg/cmd/cli/cmd/rsync/rsyncargs.go:
```diff
@@ -0,0 +1,196 @@
+				state = inEscape
+			case isSpace(c):
+				state = normal
+				args = append(args, arg.String())
```
isn't this going to append a bunch of empty (single space) arguments if i have multiple spaces between arguments, or a string like:  `"foo" "bar"`  (will treat the space between the two quoted strings as an arg based on the state transition out of "inDoubleQuote" to "inArg").

that's not strictly a problem i suppose, but it seems odd.

bparees replied:
hm. based on your tests below it seems that is not what happens but i'm not clear why.  presumably something do with how empty Buffers are handled?

csrwng replied:
umm, it's not. if you are inarg and you get a space, then you go back to normal. because you were inarg, the arg that you were building is appended to the array of args.

csrwng replied:
maybe what's throwing you off is this line:
https://github.com/openshift/origin/pull/6351/files#diff-b01b33bce1f742512c7b1e85af3f0af1R155
it's not doing anything with the space character... it's using what had already been accumulated for the current arg and adding it to the large args list.

bparees replied:
yeah i got ahead of myself, nm.

bparees commented on pkg/cmd/cli/cmd/rsync/rsyncargs_test.go:
```diff
@@ -0,0 +1,112 @@
+		},
+		{
+			input:    "--test-arg=\"my value\" --another-arg='separate value'",
+			expected: []string{"--test-arg=my value", "--another-arg=separate value"},
```
presumably `--test-arg=my value` is not what someone's going to want passed to rsync, so how are they supposed to actually specify this?  they need to double escape things so that `my value` is still quoted when it gets to rsync?

csrwng replied:
actually it is... because that's what the shell would pass  you. Remember, we're passing these as separate strings in the string array that's passed to the exec.Cmd. (they're no longer interpreted as multiple args)

mfojtik commented on pkg/cmd/cli/cmd/rsync/rsync.go:
```diff
@@ -97,6 +98,8 @@ func NewCmdRsync(name, parent string, f *clientcmd.Factory, out, errOut io.Write
 	cmd.Flags().BoolVarP(&o.Quiet, "quiet", "q", false, "Suppress non-error messages")
 	cmd.Flags().BoolVar(&o.Delete, "delete", false, "Delete files not present in source")
+
+	cmd.Flags().Var(&o.RsyncArgs, "rsync-args", "Additional arguments for rsync when using the rsync and rsync-daemon strategies.")
```
maybe add an example of how to use these?

bparees commented:
it seems like there's some support (@liggitt?  @smarterclayton?) behind just providing a minimal first class set of supported args instead of this custom parsing.

smarterclayton commented:
I'm ok with first class supported args.

On Tue, Jan 5, 2016 at 2:22 PM, Ben Parees <notifications@github.com> wrote:

> it seems like there's some support (@liggitt <https://github.com/liggitt>?
> @smarterclayton <https://github.com/smarterclayton>?) behind just
> providing a minimal first class set of supported args instead of this
> custom parsing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6351#issuecomment-169103091>.
>

csrwng commented:
Updated to use first-class flags. Also updated PR description. The flags getting exposed are the ones we know so far that there is a use case for.

@bparees

bparees commented:
@csrwng tests?  otherwise lgtm.

csrwng commented:
@bparees extended tests added
[testonlyextended][extended:core(rsync)]

bparees commented:
lgtm

csrwng commented:
[merge]
